src/workerd/io/worker.h

```diff
@@ -673,21 +673,42 @@ class Worker::Actor final: public kj::Refcounted {
   // support publicly).
   void ensureConstructed(IoContext&);
 
-  // Forces cancellation of all "background work" this actor is executing, i.e. work that is not
-  // happening on behalf of an active request. Note that this is not a part of the dtor because
-  // IoContext objects prolong the lifetime of their Actor.
+
+  // Revokes access to storage and forces cancellation of all "background work" this actor is
+  // executing, i.e. work that is not happening on behalf of an active request.
+  //
+  // To also stop execution of in-flight requests ("foreground work"), specify `mode` to be
+  // `ShutdownMode::HARD`.
   //
   // `reasonCode` is passed back to the WorkerObserver.
-  void shutdown(uint16_t reasonCode, kj::Maybe<const kj::Exception&> error = kj::none);
+  //
+  // Shutdown should only be called once.
+  //
+  // Note that this is not a part of the dtor because IoContext objects prolong the lifetime of
+  // their Actor.
+  enum class ShutdownMode {
+    // Allow in-flight requests to continue executing. They may succeed so long as they don't
+    // interact with storage, access to which is always revoked on shutdown.
+    SOFT = 0,
+    // Abort the IO context, immediately stopping execution of all in-flight requests.
+    HARD = 1
+  };
+  void shutdown(ShutdownMode mode, uint16_t reasonCode, kj::Maybe<const kj::Exception&> error = kj::none);
 
   // Stops new work on behalf of the ActorCache. This does not cancel any ongoing flushes.
   // TODO(soon) This should probably be folded into shutdown(). We'd need a piece that converts
   // `error` to `reasonCode` in workerd to do this. There may also be opportunities to streamline
   // interactions between `onAbort` and `onShutdown` promises.
   void shutdownActorCache(kj::Maybe<const kj::Exception&> error);
 
+  struct ShutdownContext {
+    // The mode passed to `shutdown()`.
+    ShutdownMode mode;
+    // The error passed to `shutdown()`.
+    kj::Maybe<kj::Exception> error;
+  };
   // Get a promise that resolves when `shutdown()` has been called.
-  kj::Promise<void> onShutdown();
+  kj::Promise<ShutdownContext> onShutdown();
```

Comment:
I guess this is a little weird now -- what if shutdown() is called multiple times, first SOFT and then HARD? Maybe we need separate onSoftShutdown() and onHardShutdown()?

Reply:
I think I'm going to shelve this PR now that we've decided we don't need to abort in-flight requests in any of the cases we discussed yesterday.

But it might be nice to revive it later since having a minimal set of shutdown modes that always leave the DO in a well-defined state regardless of what triggered the shutdown would probably make this stuff easier to reason about. And yeah I agree that, if I do revive it, I should probably handle hard shutdown coming after soft shutdown.